crates/wasi-nn/src/tf.rs

```diff
@@ -0,0 +1,406 @@
+        //   itself (`'a`) using `transmute` and use our new `self_` to borrow
+        //   the tensor data we copied to `Self`.
+        let self_ = unsafe { std::mem::transmute::<&mut Self, &'a mut Self>(self) };
+        let tensor_ref = self_.tensors.last_mut().unwrap();
+
+        use TensorTypes::*;
+        match tensor_ref {
+            TTU8(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTF16(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTF32(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTI32(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+        };
+
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        // The output requests must be made before calling session.run, or it will fail.
+        // Because we don't know which results the user will want to access,
+        // we need to save all the output tokens for later.
+
+        // Reset tokens
+        self.output_tokens.clear();
+
+        // Sort the output keys alphabetically so that we know we always index
+        // into the same key. (Note that TF's `HashMap::keys()` is returned in
+        // arbitrary order).
+        let mut output_keys: Vec<String> = self.signature.outputs().keys().cloned().collect();
+        output_keys.sort();
+
+        for i in 0..output_keys.len() {
+            let output_key = &output_keys[i as usize];
+            let output_tensor_info = self.signature.get_output(output_key)?;
+            let out_operation = self
+                .graph
+                .operation_by_name_required(&output_tensor_info.name().name)?;
+
+            // Save the output token.
+            self.output_tokens.push((
+                self.args.request_fetch(&out_operation, i as i32),
+                output_tensor_info.dtype(),
+            ));
+        }
+
+        self.session.run(&mut self.args)?;
+        Ok(())
+    }
+
+    fn get_output(&mut self, index: u32, destination: &mut [u8]) -> Result<u32, BackendError> {
+        let token_tuple = self.output_tokens[index as usize];
+        let mut results: Vec<u8> = vec![];
+
+        // Convert the output from T to u8
+        match token_tuple.1 {
+            tensorflow::DataType::UInt8 => {
+                t_to_u8_copy(&self.args.fetch::<u8>(token_tuple.0)?, &mut results);
+            }
+            tensorflow::DataType::Half | tensorflow::DataType::Float => {
+                t_to_u8_copy(&self.args.fetch::<f32>(token_tuple.0)?, &mut results);
+            }
+            tensorflow::DataType::Int32 => {
+                t_to_u8_copy(&self.args.fetch::<i32>(token_tuple.0)?, &mut results);
+            }
+            _ => {}
+        };
```

Comment:
It's unclear to me why we can't just fetch the output as a [`&[u8]`](https://github.com/abrown/wasmtime/commit/a9ddd19876a366c1509b80a623056bc0e2c0d952#diff-a940d24b7d1898c64717b16afcb125c155b4cb11526ef7e546734d2b1f17b5c0R224) and then `copy_from_slice` down below. This code does two copies!